loader/loader.go

```diff
@@ -1,5 +1,209 @@
 package loader
 
-// Loader represents loading process from a bitcoin client.
-type Loader interface {
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/IlliniBlockchain/etl-bitcoin/client"
+	"github.com/IlliniBlockchain/etl-bitcoin/database"
+	"github.com/IlliniBlockchain/etl-bitcoin/types"
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"golang.org/x/sync/errgroup"
+)
+
+// LoaderOptions represents a map of arbitrary options when constructing a LoaderManager.
+type LoaderOptions map[string]interface{}
+
+// ILoaderManager outlines an interface for a loader manager.
+type ILoaderManager interface {
+	SendInput(BlockRange, database.DBTx)
+	Close()
+}
+
+// LoaderManager stores state for managing loaders for loading data from a `Client` to
+// a `Database`.
+type LoaderManager struct {
+	client  client.Client
+	db      database.Database
+	inputCh chan *LoaderMsg[BlockRange]
+
+	ctx      context.Context
+	cancel   context.CancelFunc
+	stopOnce sync.Once
+	g        *errgroup.Group
+}
+
+// LoaderMsg stores state for data being passed through loaders.
+type LoaderMsg[T any] struct {
+	blockRange BlockRange
+	dbTx       *database.DBTx
+	data       T
+}
+
+type BlockRange struct {
+	startBlockHeight int64
+	endBlockHeight   int64
+}
+
+// NewLoaderManager creates a new LoaderManager and initiates goroutines for the loaders in a pipeline.
+func NewLoaderManager(ctx context.Context, client client.Client, db database.Database, opts LoaderOptions) (*LoaderManager, error) {
+	// initialize struct
+	ctx, cancel := context.WithCancel(ctx)
+	g, ctx := errgroup.WithContext(ctx)
+	inputCh := make(chan *LoaderMsg[BlockRange])
+	loader := &LoaderManager{
+		client:  client,
+		db:      db,
+		inputCh: inputCh,
+		ctx:     ctx,
+		cancel:  cancel,
+		g:       g,
+	}
+
+	// loaders
+	blockRangeLoader := NewLoader[BlockRange, []*chainhash.Hash](client, inputCh, blockRangeHandler)
+	g.Go(blockRangeLoader.Run)
+	blockHashLoader := NewLoader[[]*chainhash.Hash, []*types.Block](client, blockRangeLoader.Dst(), blockHashHandler)
+	g.Go(blockHashLoader.Run)
+	blockLoader := NewLoaderSink[[]*types.Block](blockHashLoader.Dst(), blockHandler)
+	g.Go(blockLoader.Run)
+
+	return loader, nil
+}
+
+// Close gracefully shuts down all loader processes.
+func (loader *LoaderManager) Close() error {
+	done := false
+	loader.stopOnce.Do(func() {
+		close(loader.inputCh)
+		loader.cancel()
+		done = true
+	})
+	if !done {
+		return fmt.Errorf("already closed")
+	}
+	return loader.g.Wait()
```

Comment:
```suggestion
	loader.stopOnce.Do(func() {
		close(loader.inputCh)
	})
	return loader.g.Wait()
```

I just realized that the context cancel here could really mess things up because after closing the first channel, the proceeding channels could still be processing things, but the context is then canceled immediately so they may not have a chance to empty out. By not canceling and just waiting we can ensure the loaders have completely finished.

Since this process is dependent on the loader terminating on their own, I think the cancelFunc is not really necessary. Doesn't really make sense now that I am looking at it (still learning how to properly utilize contexts).